plugins/inputs/modbus/request.go

```diff
@@ -42,28 +189,85 @@ func groupFieldsToRequests(fields []field, tags map[string]string, maxBatchSize
 	// For field addresses like [1, 2, 3, 5, 6, 10, 11, 12, 14] we should construct the following
 	// requests (1, 3) , (5, 2) , (10, 3), (14 , 1). Furthermore, we should respect field boundaries
 	// and the given maximum chunk sizes.
-	var requests []request
-
-	current := newRequest(fields[0], tags)
-	for _, f := range fields[1:] {
+	var groups []request
+	var current request
+	for _, f := range fields {
 		// Check if we need to interrupt the current chunk and require a new one
-		needInterrupt := f.address != current.address+current.length            // not consecutive
-		needInterrupt = needInterrupt || f.length+current.length > maxBatchSize // too large
-
-		if !needInterrupt {
+		if current.length > 0 && f.address == current.address+current.length {
 			// Still safe to add the field to the current request
 			current.length += f.length
 			if !f.omit {
-				// Omit adding the field but use it for constructing the request.
 				current.fields = append(current.fields, f)
 			}
 			continue
 		}
 
 		// Finish the current request, add it to the list and construct a new one
-		requests = append(requests, current)
-		current = newRequest(f, tags)
+		if current.length > 0 {
+			groups = append(groups, current)
+		}
+		current = newRequest(f)
+	}
+	if current.length > 0 {
+		groups = append(groups, current)
+	}
+
+	var requests []request
+	switch optimization {
+	case "shrink":
+		// Shrink request by striping leading and trailing fields with an omit flag set
+		for _, g := range groups {
+			if len(g.fields) > 0 {
+				requests = append(requests, shrinkGroup(g, maxBatchSize)...)
+			}
+		}
+	case "rearrange":
+		// Allow rearranging fields between request in order to reduce the number of touched
+		// registers while keeping the number of requests
+		for _, g := range groups {
+			if len(g.fields) > 0 {
+				requests = append(requests, optimitzeGroup(g, maxBatchSize)...)
+			}
+		}
+	case "aggressive":
+		// Allow rearranging fields similar to "rearrange" but allow mixing of groups
+		// This might reduce the number of requests at the cost of more registers being touched.
+		var total request
+		for _, g := range groups {
+			if len(g.fields) > 0 {
+				total.fields = append(total.fields, g.fields...)
+			}
+		}
+		requests = optimitzeGroup(total, maxBatchSize)
+	case "max_insert":
+		// Similar to aggressive but keeps the number of touched registers bellow a threshold
+		requests = optimitzeGroupWithinLimits(fields, maxBatchSize, maxTouchedRegisters)
```

Comment:
I think this cannot work... Can you please first collect all fields from the groups like
```suggestion
		// Similar to aggressive but keeps the number of touched registers bellow a threshold
		var total request
		for _, g := range groups {
			if len(g.fields) > 0 {
				total.fields = append(total.fields, g.fields...)
			}
		}
		requests = optimitzeGroupWithinLimits(total, maxBatchSize, maxTouchedRegisters)
```